### ISSUE TITLE
Invoke function by AddressOf Pointer

### ISSUE THREAD

**Issue by Kr00l** [WaynePhillipsEA/twinbasic#79]
**Is your feature request related to a problem? Please describe.**
VB6/VBA cannot invoke functions by a Pointer in a native way like it is done in C/C++.

**Describe the solution you'd like**
A VB-Ish syntax is needed. (shall be discussed)
Sure is that the function "template" needs to be stated in the syntax that shall be invoked.

**Describe alternatives you've considered**
Often the workaround of using CallWindowProc is used. Or better the low-level API DispCallFunc.
But if it can be avoided then it is "nice to have".

**Additional context**
Low-priority as only really needed in edge cases.

**Comment by bclothier** [comment]
FWIW, I would treat this as a request for delegate support, which was cited in #5 and alluded to in #39.

**Comment by sancarn** [comment]
@bclothier I actually disagree with that. Although invoking pointers can provide support for delegates, delegates cannot provide support for invoking all pointers.

See [Universal DLL Calls](https://www.vbforums.com/showthread.php?781595-VB6-Call-Functions-By-Pointer-(Universall-DLL-Calls)) for examples of invoking pointers. I feel a bigger feature request here is:

**Ability to invoke any pointer with any calling convention**

```vb
Dim ptr as LongPtr
ptr = AddressOf myFunc
retVar = ptr.stdInvoke(argtypes, argvalues)
ptr = getCFunctionPointer()
ptr.invoke(CC_CDECL, argtypes, argvalues)
ptr = objptr(something)
ptr.invokeObject(vtableoffset, argtypes, argvalues)
```

For information - call conventions it'd be great to support:

```vb
Public Enum CALLINGCONVENTION_ENUM
  ' http://msdn.microsoft.com/en-us/library/system.runtime.interopservices.comtypes.callconv%28v=vs.110%29.aspx
  CC_FASTCALL = 0&
  CC_CDECL
  CC_PASCAL
  CC_MACPASCAL
  CC_STDCALL                        ' typical windows APIs
  CC_FPFASTCALL
  CC_SYSCALL
  CC_MPWCDECL
  CC_MPWPASCAL
End Enum
```

**Comment by bclothier** [comment]
[In C#, you can specify the calling conventions for your delegates](https://stackoverflow.com/questions/5155180/changing-a-c-sharp-delegates-calling-convention-to-cdecl).

**Comment by wqweto** [comment]
Standard pointer-to-function delegates look very similar to an API function declare as syntax. It's an API function with user-provided pfn to call to while the API declare has to keep internal book-keeping about retrieving hModule and pfn on first call.

Member pointer-to-function delegates are a different beast altogether.

**Comment by Kr00l** [comment]
Since I started this feature request I suggest my "syntax" proposal.

Like in a API declare, but instead of the "Lib" keyword the keyword "CallByAddressOf" is used. It's a keyword to declare a function "prototype" with it's signature only.

Those "APIs" with the keyword CallByAddressOf needs in it's first argument "automagically" the pfn. (AddressOf, "ProcPtr")

Example A: (invoke dll func pointer)
```
Private Declare Function MyGetDC CallByAddressOf ( ByVal hWnd As Long) As Long

Dim pfn As Long
pfn = GetProcAddress(LoadLibrary("user32"), "GetDC")
Dim hDC As Long
hDC = MyGetDC(pfn, hWnd)
```

Example B: (invoke VB address pointer)
```
Private Declare Sub MyVBSub CallByAddressOf ( ByRef szText As String, ByVal szAdd As String) As Long

Dim pfn As Long
pfn = ProcPtr(AddressOf VBConcat)
MyVBConcat pfn, Text1, Text2

Private Function ProcPtr(ByVal Addr As Long) As Long
ProcPtr = Addr
End Function

Private Sub VBSub(ByRef szText As String, ByVal szAdd As string)
szText = szText & szAdd
End Sub
```

**Comment by vbRichClient** [comment]
The 'CallByAddressOf' meaning is already contained in the 'Declare' Keyword.

E.g. in FreeBasic your Example A could be written:
```
Private Declare Function MyGetDC (ByVal hWnd As Long) As Long

Dim SomeHdc As Long
If MyGetDC = 0 Then MyGetDC = DylibSymbol(DylibLoad("user32"), "GetDC")
SomeHdc = MyGetDC(hWnd)
```
(DylibLoad, DylibSymbol, DylibFree being runtime-functions, which abstract from the underlying system-APIs)

And your example B could be written this way:
```
Private Declare Sub MyVBSub (ByRef szText As String, ByVal szAdd As String)
 
If MyVBSub = 0 Then MyVBSub = AddressOf VBSub '<- the compiler recognizes what is meant, due to the AddressOf Keyword
MyVBSub Text1, Text2

Private Sub VBSub(ByRef szText As String, ByVal szAdd As string)
    szText = szText & szAdd
End Sub
```
There's also recognition at DIM-level for Function-Signature-Defs, like:
Dim MyVBSub As Sub(ByRef szText As String, ByVal szAdd As String)
(the above is working in a similar way also for Function-Defs in UDTs) 

There's another thing I found immediately "nice to have" when I played with FreeBasic.
In addition to "right-hand-side" Type-specifying, it allows to define Vars also "left-typed":
`Dim As Long i`
Especially nice with that is, that now a sequence of vars with the same type can be detected unambigously, when written thus:
`Dim As Long i, j, k  `
Or within a UDT:
```
Type t3DPtDbl
   As Double x, y, z, w
End Type
```
I'd suggest - when new *Syntax* (not existing in VB6) is considered for some new feature in twinBasic, 
to first look for "prior art" in the following "compatibility-order":
- VBA64  (as e.g. LongPtr-specifiers)
- FreeBasic (e.g. the cdecl-support, but also AndAlso, OrElse of twinBasic are already compatible)
- VB.NET 

Olaf

**Comment by Kr00l** [comment]
Thanks Olaf for your input. Yes, your syntax looks cleaner by omitting 'CallByAddressOf' and of course omit the 'Lib' keyword. So just a blank 'Declare' would be sufficient.

Also to assign the 'pfn' like a 'property let' is cool, since then it needs to be assigned only once.
I was not aware of this "prior art" as I don't use FreeBasic. But is I think/hope that Wayne is aware to lookout of these. :-)

**Comment by bclothier** [comment]
I am not sure how I feel about the `AddressOf` being modified to allow assignment to a `LongPtr` variable/argument. That seems to open an avenue for accidental implicit conversion, wouldn't it? 

As mentioned elsewhere, I am not a fan of implicit conversions VB* likes to do as that usually hides potential bugs that becomes runtime errors. In .NET delegates are not equivalent to a function pointer and cannot be assigned as such except across a p/invoke and usually needs `MarshalAs(UnmanagedType.FunctionPointer)` IIRC. The good thing is that we never need to do `if(ptr) ptr(...)`; we just `ptr.Invoke(...)` and all the housekeeping stuff is tidily abstracted out.

Regarding delegates, one other thing I want to be sure to consider is that they will be treated as a type, and not as some kind of statement. Treating them as a type allows reuse and sharing and would also allow for generics support as well.

**Comment by WaynePhillipsEA** [comment]
@bclothier  Yes, delegates will be treated as a typed function pointer, internally treated as a type.

We will have to support coercing to LongPtr for backwards compatibility.   We can definitely restrict that when Option Strict is on though!

**Comment by mansellan** [comment]
should this be moved to lang-design and converted to a discussion?
